### DISCUSSION TITLE
Different outputs when dumping raw vs using both --format and --raw parameters.

### DISCUSSION THREAD

**Discussion by Digitoxin1** [keirf/greaseweazle#396] [Q&A]
I'm getting different outputs when dumping raw vs when using both the --format and --raw parameters together.

I'm using a 720K floppy disk with copy protection on track 6, side 0.  
When I dump a disk using the command: `gw read track00.0.raw` my raw tracks look like the first image.
When I dump using the command `gw read --format ibm.720 --raw track00.0.raw` my raw tracks look like the second image.
Notice there is a huge file size disparity between the file sizes of the tracks, especially on track06.0.raw, which is the copy protected track.

What is Greaseweazle doing differently when dumping the first way vs the second and how does that affect my .raw track outputs?  Why are the track file sizes so small when using --format and --raw together, but the one copy protected track is significantly larger?

![Screenshot 2024-01-10 013011](https://github.com/keirf/greaseweazle/assets/5696659/c58a1929-4fb7-4b87-9176-cce218759d20)
![Screenshot 2024-01-10 013040](https://github.com/keirf/greaseweazle/assets/5696659/dec2b317-2d0e-4437-8ab0-49cb0e8e89a4)

**Comment by keirf** [comment]
With no format specified, gw read defaults to reading three revolutions per track. When a format is specified, this default is overridden by the relevant codec. For IBM track formats the default is two revolutions. Hence your stream files are 2/3 the size in this case.

The number of revolutions is reduced because the codec can detect errors and retry as necessary. This is what happens in the copy protection track. It appears malformed compared with what the codex expects, so further revolutions are read from disk. These get concatenated together in the raw output stream file. Hence it is much larger.

**Comment by Digitoxin1** [comment]
Okay, I may have found a bug then.  I redumped the disk using the command:
`gw read --revs 3 --format ibm.720 --raw track00.0.raw`

All tracks now dump at 3 revolutions except for the copy protected track 6.0, which is still dumping at 409 KB, ignoring the number of revolutions I requested.

**Reply by keirf** [reply]
Retries always get concatenated. So revs ends up being a minimum. If you don't want retries then set `--retries=0`

**Reply by Digitoxin1** [reply]
How are retries stored in the .raw vs revs or is there no functional difference?

**Reply by keirf** [reply]
There's usually no way to discriminate so they simply get concatenated. Hence there will be discontinuities at the splices (which occur at index points). For Kryoflux streams arguably we could put retries in their own files with a naming convention.

So yes they look like extra revs.

**Reply by Digitoxin1** [reply]
So, If I did the following:

`gw read --revs 3 --retries=3 --format ibm.720 --raw track00.0.raw`

It would limit me to up to three retries per track and will do three revs per track, so my resulting output will have three revs on every track similar to if I just did:

`gw read track00.0.raw`

Or, will I end up with 9 revs on track 6 because it would do three retries of three revs each?

**Reply by keirf** [reply]
You'll end up with 12 revs on track 6. Original 3 plus 3*3 retries.

**Reply by Digitoxin1** [reply]
Got it.  Thanks for the clarification.
